packages/core/src/browser/preferences/preference-service.ts

```diff
@@ -43,29 +69,78 @@ export interface PreferenceChanges {
+        return this.change.scope;
+    }
+    get domain() {
+        return this.change.domain;
+    }
+    get newValue() {
+        return this.change.newValue;
+    }
+    get oldValue() {
+        return this.change.oldValue;
+    }
+
+    canAffect(resourceUri?: string): boolean {
+        if (this.domain && resourceUri &&
+            this.domain.length !== 0 &&
+            this.domain.map(uriStr => new URI(uriStr))
+                .every(folderUri => folderUri.path.relativity(new URI(resourceUri).path) < 0)
+        ) {
+            return false;
+        }
+        for (const [scope, provider] of this.providers.entries()) {
+            if (!resourceUri && scope === PreferenceScope.FOLDER) {
+                continue;
+            }
+            const providerInfo = provider.canProvide(this.preferenceName, resourceUri);
+            const priority = providerInfo.priority;
+            if (priority >= 0 && scope > this.scope) {
+                return false;
+            }
+            if (scope === this.scope && this.domain.some(d => providerInfo.provider.getDomain().findIndex(pd => pd === d) < 0)) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
 export const PreferenceService = Symbol('PreferenceService');
 export interface PreferenceService extends Disposable {
     readonly ready: Promise<void>;
     get<T>(preferenceName: string): T | undefined;
     get<T>(preferenceName: string, defaultValue: T): T;
-    get<T>(preferenceName: string, defaultValue?: T): T | undefined;
-    set(preferenceName: string, value: any, scope?: PreferenceScope): Promise<void>;
-    onPreferenceChanged: Event<PreferenceChange>;
+    get<T>(preferenceName: string, defaultValue: T, resourceUri: string): T;
+    get<T>(preferenceName: string, defaultValue?: T, resourceUri?: string): T | undefined;
+    set(preferenceName: string, value: any, scope?: PreferenceScope, resourceUri?: string): Promise<void>;
+    onPreferenceChanged: Event<PreferenceDataChange>;
 }
 
 /**
  * We cannot load providers directly in the case if they depend on `PreferenceService` somehow.
  * It allows to load them lazilly after DI is configured.
  */
 export const PreferenceProviderProvider = Symbol('PreferenceProviderProvider');
-export type PreferenceProviderProvider = (scope: PreferenceScope) => PreferenceProvider;
+export type PreferenceProviderProvider = (scope: PreferenceScope, uri?: URI) => PreferenceProvider;
 
 @injectable()
 export class PreferenceServiceImpl implements PreferenceService, FrontendApplicationContribution {
 
     protected preferences: { [key: string]: any } = {};
 
-    protected readonly onPreferenceChangedEmitter = new Emitter<PreferenceChange>();
+    protected readonly onPreferenceChangedEmitter = new Emitter<PreferenceDataChange>();
     readonly onPreferenceChanged = this.onPreferenceChangedEmitter.event;
 
     protected readonly onPreferencesChangedEmitter = new Emitter<PreferenceChanges>();
```

Comment:
it does not seem to be used, but there are client using it, please fire events properly

Reply:
you are right. i removed this event and updated the client.